src/Data/Bin/Properties.agda

```diff
@@ -1,143 +1,913 @@
+
+fromℕ∘pred :  (n : ℕ) → fromℕ (predₙ n) ≡ pred (fromℕ n)
+fromℕ∘pred n =
+  let (x , toℕ-x≡n) = ext-fromℕ n
+  in
+  begin
+    fromℕ (predₙ n)         ≡⟨ cong (fromℕ ∘ predₙ) (sym (toℕ∘fromℕ n)) ⟩
+    fromℕ (predₙ (toℕ x))   ≡⟨ cong fromℕ (sym (toℕ∘pred x)) ⟩
+    fromℕ (toℕ (pred x))    ≡⟨ fromℕ∘toℕ (pred x) ⟩
+    pred x                  ≡⟨ refl ⟩
+    pred (fromℕ n)
+  ∎
+
+
+------------------------------------------------------------------------------
+-- Addition properties. Chapter II.
+-- Commutativity and associativity for _+_ are proved by using the isomorphism
+-- to ℕ.
+------------------------------------------------------------------------------
+
++-comm :  Commutative _+_
++-comm a b = begin
+  a + b                    ≡⟨ sym (fromℕ∘toℕ (a + b)) ⟩
+  fromℕ (toℕ (a + b))      ≡⟨ cong fromℕ (toℕ+homo a b) ⟩
+  fromℕ (toℕ a +ₙ toℕ b)   ≡⟨ cong fromℕ (ℕp.+-comm (toℕ a) (toℕ b)) ⟩
+  fromℕ (toℕ b +ₙ toℕ a)   ≡⟨ cong fromℕ (sym (toℕ+homo b a)) ⟩
+  fromℕ (toℕ (b + a))      ≡⟨ fromℕ∘toℕ (b + a) ⟩
+  b + a                    ∎
+
++-assoc :  Associative _+_
++-assoc a b c = begin
+  (a + b) + c                  ≡⟨ sym (fromℕ∘toℕ ((a + b) + c)) ⟩
+  fromℕ (toℕ ((a + b) + c))    ≡⟨ cong fromℕ (toℕ+homo (a + b) c) ⟩
+  fromℕ (toℕ (a + b) +ₙ cN)    ≡⟨ cong (fromℕ ∘ (_+ₙ cN)) (toℕ+homo a b) ⟩
+  fromℕ ((aN +ₙ bN) +ₙ cN)     ≡⟨ cong fromℕ (ℕp.+-assoc aN bN cN) ⟩
+  fromℕ (aN +ₙ (bN +ₙ cN))     ≡⟨ cong (fromℕ ∘ (aN +ₙ_))
+                                                (sym (toℕ+homo b c)) ⟩
+  fromℕ (aN +ₙ toℕ (b + c))    ≡⟨ cong fromℕ (sym (toℕ+homo a (b + c))) ⟩
+  fromℕ (toℕ (a + (b + c)))    ≡⟨ fromℕ∘toℕ (a + (b + c)) ⟩
+  (a + (b + c))                ∎
+  where
+  aN = toℕ a;   bN = toℕ b;   cN = toℕ c
+
+
++-cancelˡ-≡ : LeftCancellative _+_       -- the proof is via toℕ
++-cancelˡ-≡ x {y} {z} x+y≡x+z = begin
+  y         ≡⟨ sym (fromℕ∘toℕ y) ⟩
+  fromℕ m   ≡⟨ cong fromℕ m≡n ⟩
+  fromℕ n   ≡⟨ fromℕ∘toℕ z ⟩
+  z         ∎
+  where
+  k = toℕ x;  m = toℕ y;  n = toℕ z
+
+  eq = begin
+    k +ₙ m         ≡⟨ sym (toℕ+homo x y) ⟩
+    toℕ (x + y)    ≡⟨ cong toℕ x+y≡x+z ⟩
+    toℕ (x + z)    ≡⟨ toℕ+homo x z ⟩
+    k +ₙ n         ∎
+
+  m≡n = begin
+    m                ≡⟨ sym (ℕp.m+n∸n≡m m k)⟩
+    (m +ₙ k) ∸ₙ k    ≡⟨ cong (_∸ₙ k) (ℕp.+-comm m k) ⟩
+    (k +ₙ m) ∸ₙ k    ≡⟨ cong (_∸ₙ k) eq ⟩
+    (k +ₙ n) ∸ₙ k    ≡⟨ cong (_∸ₙ k) (ℕp.+-comm k n) ⟩
+    (n +ₙ k) ∸ₙ k    ≡⟨ ℕp.m+n∸n≡m n k ⟩
+    n                ∎
+
++-cancelʳ-≡ : RightCancellative _+_
++-cancelʳ-≡ {x} y z y+x≡z+x =  +-cancelˡ-≡ x {y} {z} x+y≡x+z
+  where
+  x+y≡x+z = begin
+    x + y   ≡⟨ +-comm x y ⟩
+    y + x   ≡⟨ y+x≡z+x ⟩
+    z + x   ≡⟨ +-comm z x ⟩
+    x + z   ∎
+
+decEquivalence :  IsDecEquivalence (_≡_ {A = Bin})
+decEquivalence =  record{ isEquivalence = isEquivalence;  _≟_ = _≟_ }
+
+≡-decSetoid : DecSetoid 0ℓ 0ℓ
+≡-decSetoid = record
+  { Carrier          = Bin
+  ; _≈_              = _≡_
+  ; isDecEquivalence = decEquivalence
```

Comment:
No need to provide the carrier and equality explicitly here.

Reply:
Removed these unneeded things.